### PR TITLE
EG: SlideGraph Pipeline As Advance Example Usage

### PR DESCRIPTION
**Pull request by vqdang**
Finally something to bind all the parts in the toolbox together for a single purpose.

- [x] Framework conversion
- [x] Inference integrity check
- [x] Training integrity check
- [x] Add cell compositions
- [x] Explanation  
- [ ] ~~Performance Tests~~

There are several parts / considerations that may need to be split into / implemented by other PR during the making of this notebook to reduce its complexity

- Generating some specific train/valid/test split
- Drawing generic graph
- Drawing graph by a specific method
- Classifcal feature extraction
  - Cell compositions of patches
  - Nuclei morphology
- Extending some sklearn with `partial_fit` as fitting 1mllion ResNet50 vectors is impossible (OOM)
- Full training engine (may be far into the future)

Lastly, full pipeline as a notebook is bad for organizing code. Perhaps, we should maintain a subfolder called `pipeline` within `examples`, such as `examples/pipeline/slidehgraph` will contain `.py` code of slide graph (parts split/merge from notebook). Additionally, due to the complexity of making a pipeline, tests should also be written, however, notebooks won't facilitate that easily. These tests will focus solely on behavior tests i.e at least being able to replicate the original paper results.

### CODE REVIEW

File: examples/slide_graph.ipynb [hunk 1]
```diff
@@ -0,0 +1 @@
...sr90RI6qNbjtX2Z73ie7NVFQ9IWaViWytYbwqKSCCOKsiBY3zWmzrmqCLPjRRGbdONSi2v0SI6EBkbWptX9QZWgfEdOCwnaWaevJdzbGMWa+WtQnbeIRfKx2atp0vjvTvEVxwOj9lRAcmq3Jea9UOsS4B3S+m/7CVG74Ks5/N6rpEnhpOmbM2/OX4wY9WCmWJlDle8solr1zyyh9yXvHWM4aBm+0V7z99D2MM++OBw3x4880N+txr4bAcef7wko2faNVQq8blS175g8srb14cdFOTVgo1a6eErINWLWbSiwdynzlYuyrL/ZF4ms/cSHpQi3V1cO3mDrFw+uyBAt3Nz3dXwMC4Uek0GQPiLS1XyhypqSBNVRFEhDB6/KSawNWqC12pEGOmoTd5MyrIU6vT7kNO7EaFNk/9BcqIdDjUsR2Cau06y9jcmes2TQMheFpTt83WdDDNe60Wc9/8S8y8fDiyxIq3FmNUes07TxOLGMtgFborpZLaamil8mZqJV9JpeGcckyts9gpsKtbTQrGgbOYAMNuwA2OaCotR1Zds5XLOTivQ1TTQNiNGGtUyWaFJmc1sBHU7l6MMN5sGJ7umOfI3WnhlAvNCKeWsbUypcx2zj1nLECjiVA67DcfI6UURBq3VxNXm0FN84DWIffl1RE/ea4/fcLuvR02eML1Fhs8GJT/K/AYxhslVdKifNP7+aSKBL4SWsM7x5VzTMPA1TQyDgODdzQxNKcUAyla5NamsoI1aWclVQ3lg7UEY3Ug7BjJqWIGjx0DxjaWBg+x4G3Dx6wdh01DgsUGlfItVQNHEUs1Qi2asGqtHE+Rw3FRLeXjkRiXnhiUs3jlPLfBYUV1jb0xUJp2Iou6JuZcsCLsH07YBttcuH0vAUU5v1apAMU6kMwq0WioeO/Y7TbUqgNvzqvquaq+FJwIm2HAimqah+B7B0SRBYCHY0RQybdjdwC9vXYMwas4wdWIn4IqxiwaM6RUTMuU2ojxsROaij7bUhs5VZwVtlTaEinOko+L6lCXejYJKinpsFbVZ451WtT04qXm3DmyFTHKW1YTHQ2s0zhw1eOL6/4RbhsYbrcgMH9yx/zpnTppLpmYKn7wbHdKw5hj1v0vXeXF6PnHlMkp492An0acMYTB4n2XgLwcP/BRjgunj15yONlLXrnklUte+SPJKwM/9uU/w//2v/gHvDo
```
Comment (vqdang) [inline examples/slide_graph.ipynb:1]: <p>Sample comment</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TissueImageAnalytics/tiatoolbox/pull/180/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Review by DavidBAEpstein — CHANGES_REQUESTED** [review]
[sg_edited.docx](https://github.com/TissueImageAnalytics/tiatoolbox/files/7542092/sg_edited.docx)
I suggest accepting or rejecting my suggested changes using the facilities of Word.
Subsequently one can copy and paste chunks of text into the notebook. Note that the conversion process ipynb -> docx does not deal correctly with code snippets in text cells or with python variable names in text cells. So some detailed manual work may be needed.

I have put some important remarks into comments. Please read them.

File: tiatoolbox/tools/scale.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
+import numpy as np
```
Comment (shaneahmed) [inline tiatoolbox/tools/scale.py:1]: licensing information missing at the top.

File: tiatoolbox/utils/visualization.py [hunk 3]
```diff
@@ -224,3 +219,58 @@ def overlay_prediction_contours(
...
+            M is the number of edges. Each edge is defined as `(src, dst)`
+            where each is respectively the index of within `nodes`.
+        node_colors (tuple or np.ndarray): A color or list of node colors.
+            Each color is expected to be `(r, g, b)` and is between 0-255.
+        edge_colors (tuple or np.ndarray): A color or list of node colors.
+            Each color is expected to be `(r, g, b)` and is between 0-255.
+        node_size (int): Radius of each node.
+        edge_size (int): Linewidth of the edge.
+
+    """
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:253]: can we add example usage?

**Review by DavidBAEpstein — COMMENTED** [review]
Here are my comments of slide_graph.ipynb. Comments on the inference notebook will come later. I changed the format ipynb -> html -> docx, so that I could make suggestions in Track Mode. I attach the docx file. Please find my suggestions by using review mode in Word and clicking on "Next" for the next suggestion. Also please look at the comments in the comment boxes. If you are puzzled by any particular suggestion, I would be happy to explain my reasons.
[DE_TRK_slide_graph.docx](https://github.com/TissueImageAnalytics/tiatoolbox/files/7637027/DE_TRK_slide_graph.docx)
Comments on slide_graph_inference.ipynb still to come..

**Review by DavidBAEpstein — CHANGES_REQUESTED** [review]
Small suggestions are easy to miss. Please use TRK next/previous Accept/Reject to find all my suggestions. I try to remove words or phrases that do not add information. In general, shorter sentences are easier to understand. Please check my suggestions to make sure they are correct.
[DE_TRK_slide_graph_inference.docx](https://github.com/TissueImageAnalytics/tiatoolbox/files/7649516/DE_TRK_slide_graph_inference.docx)

**Review by DavidBAEpstein — COMMENTED** [review]
I went through examples/inference-pipelines/slide-graph.ipynb and made a few (very few) suggestions. I decided not to bother with Colab, as this has been checked by Dang and probably by the other reviewers. I will also look at the full pipeline version for grammar and meaning.
[2DE-TRK-slide-graph.docx](https://github.com/TissueImageAnalytics/tiatoolbox/files/7668318/2DE-TRK-slide-graph.docx)
.

**Review by DavidBAEpstein — COMMENTED** [review]
Find suggested changes using Track Changes
[DE-TRK-full-pipeline-slide-graph.docx](https://github.com/TissueImageAnalytics/tiatoolbox/files/7678780/DE-TRK-full-pipeline-slide-graph.docx)
.

**Pull request merged**